EVAKA-4190: Citizen sends a new message

mapehe opened the pull request:
#### Summary

Citizen is able to send a new message.

kimifoud commented on frontend/src/citizen-frontend/messages/MessagesPage.tsx:
```diff
@@ -35,6 +44,16 @@ export default React.memo(function MessagesPage() {
+    Loading.of()
+  )
+  const loadReceivers = useRestApi(getReceivers, setReceivers)
+  const { refreshThreads } = useContext(MessageContext)
```
`useContext(MessageContext)` is already defined above.

kimifoud commented on frontend/src/citizen-frontend/messages/MessagesPage.tsx:
```diff
@@ -35,6 +44,16 @@ export default React.memo(function MessagesPage() {
+  const loadReceivers = useRestApi(getReceivers, setReceivers)
+  const { refreshThreads } = useContext(MessageContext)
+
+  useEffect(() => {
+    loadReceivers()
+  }, [loadReceivers])
```
```suggestion
  useEffect(() => loadReceivers(), [loadReceivers])
```

kimifoud commented on frontend/src/citizen-frontend/messages/MessageEditor.tsx:
```diff
@@ -0,0 +1,176 @@
+})
+
+const Container = styled.div`
+  width: 680px;
```
Shouldn't we design the citizen frontend mobile first?

mapehe commented:
Thanks for the detailed feedback @kimifoud 💎 

Implemented most of your suggestions and added both integration test for citizen query and e2e tests for citizen sending a message.

kimifoud commented on frontend/src/citizen-frontend/messages/MessagesPage.tsx:
```diff
@@ -58,6 +80,19 @@ export default React.memo(function MessagesPage() {
+          onSend={(message: SendMessageParams) =>
+            sendMessage(message).then(() => {
+              setEditorVisible(false)
+              refreshThreads()
```
This needs to be addressed.
If errors are not handled (res.isFailure), it just reloads the page and the message is lost.
Also, double clicking sends two messages.

kimifoud commented on frontend/src/citizen-frontend/messages/state.tsx:
```diff
@@ -119,6 +121,11 @@ export const MessageContextProvider = React.memo(
 
     const loadMessages = useRestApi(getReceivedMessages, setMessagesResult)
+    const refreshThreads = useCallback(() => {
+      setThreads({ ...initialThreadState, currentPage: 0 })
```
`initialThreadState` already has `currentPage: 0`

kimifoud commented on frontend/src/e2e-playwright/pages/citizen/citizen-messages.ts:
```diff
@@ -18,9 +24,39 @@ export default class CitizenMessagesPage {
+
+  async sendNewMessage(title: string, content: string, receivers: string[]) {
+    await this.#newMessageButton.click()
+    await waitUntilEqual(() => this.isEditorVisible(), true)
+    await this.#inputTitle.fill(title)
+    await this.#inputContent.fill(content)
+    await this.#receiverSelection.click()
+    for (const receiver of receivers) {
+      await this.page.click(`text="${receiver}"`)
+    }
+    await this.#sendMessageButton.click()
+    await waitUntilEqual(
```
`waitUntilTrue` could be used.

kimifoud commented on service/src/integrationTest/kotlin/fi/espoo/evaka/messaging/MessageQueriesTest.kt:
```diff
@@ -323,6 +347,67 @@ class MessageQueriesTest : PureJdbiTest() {
     }
 
+    @Test
+    fun `asdasdasd`() {
```
😁

kimifoud commented on service/src/integrationTest/kotlin/fi/espoo/evaka/messaging/MessageQueriesTest.kt:
```diff
@@ -323,6 +347,67 @@ class MessageQueriesTest : PureJdbiTest() {
+        }
+        val supervisorPersonalAccount = db.read {
+            it.getEmployeeMessageAccounts(employee1Id)
+        }.filter { it != group1Account && it != group2Account }.first()
```
Could be simplified
```suggestion
        }.first { it != group1Account && it != group2Account }
```

kimifoud commented on service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageControllerCitizen.kt:
```diff
@@ -79,6 +86,16 @@ class MessageControllerCitizen(
+        db: Database.Connection,
+        user: AuthenticatedUser,
+    ): List<MessageAccount> {
+        Audit.MessagingSentMessagesRead.log()
```
Wrong audit log type.

kimifoud commented on service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageControllerCitizen.kt:
```diff
@@ -100,6 +117,32 @@ class MessageControllerCitizen(
+        user: AuthenticatedUser,
+        @RequestBody body: CitizenMessageBody,
+    ): List<UUID> {
+        val accountId = requireMessageAccountAccess(db, user)
```
Audit logging is missing

kimifoud commented on service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageQueries.kt:
```diff
@@ -270,6 +271,64 @@ fun Database.Read.getMessage(id: UUID): Message {
+             AS (SELECT gplt.daycare_group_id AS id
+                 FROM   placement_ids
+                        INNER JOIN daycare_group_placement gplt
+                                ON placement_ids.id = gplt.daycare_placement_id) SELECT
+        msg.id                                 AS id,
+        msg_name.account_name AS name
+        FROM   supervisors
+               INNER JOIN employee e
```
Errr, the formatting is still quite hairy 😬

kimifoud commented on frontend/src/citizen-frontend/messages/MessageEditor.tsx:
```diff
@@ -0,0 +1,177 @@
+interface Props {
+  receiverOptions: MessageAccount[]
+  onSend: (messageBody: SendMessageParams) => void
+  onClose: () => void
+  onDiscard: () => void
```
You can drop `onDiscard` and just re-use `onClose`

kimifoud commented on frontend/src/citizen-frontend/messages/MessagesPage.tsx:
```diff
@@ -58,6 +82,23 @@ export default React.memo(function MessagesPage() {
+                setDisplaySendError(false)
+                refreshThreads()
+              })
+              .catch(() => setDisplaySendError(true))
```
This `catch` is never reached because all exceptions are mapped to `Failure` in `sendMessage`. The failure state needs to be handled in `.then`

kimifoud commented on frontend/src/lib-customizations/espoo/citizen/assets/i18n/sv.tsx:
```diff
@@ -138,6 +138,17 @@ const sv: Translations = {
       BULLETIN: 'Bulletin'
+    },
+    messageEditor: {
+      newMessage: 'Nytt Meddelande',
+      receivers: 'Mottagare',
+      title: 'Titel',
+      message: 'Meddelande',
+      deleteDraft: 'Ta bort utkast',
+      send: 'Skicka',
+      search: 'Sök',
+      noResults: 'Inga resultat',
+      messageSendError: 'Misslyckades med att skicka meddelande'
```
So have these been checked with someone who knows the language?

mapehe replied:
Usually translations have been done by someone more knowledgeable only when there is a lot of text. For example, when the new application editor was built.

kimifoud commented on service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageControllerCitizen.kt:
```diff
@@ -79,6 +86,16 @@ class MessageControllerCitizen(
+        user: AuthenticatedUser,
+    ): List<MessageAccount> {
+        val accountId = requireMessageAccountAccess(db, user)
+        Audit.MessagingCitizenFetchReceiversForAccount.log(targetId = accountId)
```
Audit logging needs to happen first, before `requireMessageAccountAccess`. Otherwise some requests are not logged if errors are thrown from the function that is called first.

kimifoud commented on service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageControllerCitizen.kt:
```diff
@@ -100,6 +117,33 @@ class MessageControllerCitizen(
+        @RequestBody body: CitizenMessageBody,
+    ): List<UUID> {
+        val accountId = requireMessageAccountAccess(db, user)
+        Audit.MessagingCitizenSendMessage.log()
```
Same here.

kimifoud commented on service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageQueries.kt:
```diff
@@ -270,6 +271,74 @@ fun Database.Read.getMessage(id: UUID): Message {
+    FROM guardian g
+    JOIN placement pl
+    ON g.child_id = pl.child_id
+    WHERE guardian_id = (
+        SELECT person_id AS id
+        FROM message_account
+        WHERE id = :accountId
+    )
```
This sub query can be removed if you pass the citizen's id from the controller.

Joosakur commented on service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageQueries.kt:
```diff
@@ -272,6 +273,74 @@ fun Database.Read.getMessage(id: UUID): Message {
+    // language=SQL
+    val sql = """
+WITH placement_ids AS (
+    SELECT pl.id AS id
+    FROM guardian g
```
Now we take children whose guardian the person is, but I assume that the head of child should also be allowed to send a message even without guardianship (e.g. case sijoitusperhe). On the other hand you could ask whether we should exclude guardians who have been blocklisted on child's page (messages won't go to them but are they still allowed to send new messages?)

mapehe replied:
I updated the logic:

* Fridge parent has the right to send a new message similarly to a guardian.
* A blocklisted recipient has no right to send a new message. (Also created a test that checks that this works as intended)

Joosakur commented on service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageQueries.kt:
```diff
@@ -272,6 +273,74 @@ fun Database.Read.getMessage(id: UUID): Message {
+    AND Daterange(pl.start_date, pl.end_date, '[]') @> current_date
+),
+supervisors AS (
+    SELECT e.id AS id
```
maybe `SELECT DISTINCT`

Joosakur commented on service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageQueries.kt:
```diff
@@ -272,6 +273,74 @@ fun Database.Read.getMessage(id: UUID): Message {
+    WHERE acl.role = 'UNIT_SUPERVISOR'
+),
+groups AS (
+    SELECT gplt.daycare_group_id AS id
```
maybe `SELECT DISTINCT`

Joosakur commented on service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageQueries.kt:
```diff
@@ -272,6 +273,74 @@ fun Database.Read.getMessage(id: UUID): Message {
+        FROM message_account
+        WHERE id = :accountId
+    )
+    AND Daterange(pl.start_date, pl.end_date, '[]') @> current_date
```
probably ok for this PR, but I anticipate a need to contact at least the supervisor in advance some time before the placement starts

Joosakur commented on service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageQueries.kt:
```diff
@@ -170,7 +170,8 @@ threads AS (
             FROM participated_messages rec
-            WHERE rec.thread_id = t.id AND rec.recipient_id = :accountId)
+            WHERE rec.thread_id = t.id
+            AND (rec.recipient_id = :accountId OR ${if (includeWhereSender) "rec.sender_id = :accountId" else "false"}))
```
I'm not sure about this change. This may also break current employee side so that sent messages are shown in inbox

mapehe replied:
Thanks for the feedback. Good point.

The parameter `includeWhereSender` was originally called something like `notEmployee`. It's intended to always be `false` when making an employee-side queries. This was later renamed as suggested by the other reviewer. Maybe `notEmployee` would indeed be more legible?

mapehe replied:
Renamed the parameter, since in my opinion, it clarifies the context a bit.

mapehe commented:
Thanks for the feedback @Joosakur. Made changes accordingly. Also went through the notes from our call and implemented several small changes based on that discussion.

Joosakur commented on service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageControllerCitizen.kt:
```diff
@@ -125,22 +126,28 @@ class MessageControllerCitizen(
-                tx.insertRecipients(setOf(it.id), messageId)
-                threadId
+        val validReceivers = db.read { it.getCitizenReceivers(accountId) }
+        val allReceiversValid = !body.recipients.map { validReceivers.contains(it) }.contains(false)
```
I believe this could be written as `body.recipients.all { validReceivers.contains(it) }`

Joosakur commented on service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageQueries.kt:
```diff
@@ -291,9 +291,38 @@ WITH placement_ids AS (
+        FROM message_account
+        WHERE id = :accountId
+    )
+    AND Daterange(pl.start_date, pl.end_date, '[]') @> current_date
```
should also have `AND Daterange(fg.start_date, fg.end_date, '[]') @> current_date AND fg.conflict = false`
